### PR TITLE
Add -ftrapv to CFLAGS and CXXFLAGS when --enable-debug is used. Enable -ftrapv in Travis.

### PR DESCRIPTION
**Pull request by practicalswift**
By generating a trap for signed overflow on addition, subtraction, multiplication operations in the Travis testing we are more likely to identify problematic code prior to merging it.

### CODE REVIEW

**Comment by fanquake** [comment]
```
The command "test -n "$USE_SHELL" && eval '"$USE_SHELL" -c "./autogen.sh"' || ./autogen.sh" exited with 0.
$ mkdir build && cd build

The command "mkdir build && cd build" exited with 0.
$ ../configure --cache-file=config.cache $BITCOIN_CONFIG_ALL $BITCOIN_CONFIG || ( cat config.log && false)
configure: error: unrecognized option: `-ftrapv''
Try `../configure --help' for more information
cat: config.log: No such file or directory

The command "../configure --cache-file=config.cache $BITCOIN_CONFIG_ALL $BITCOIN_CONFIG || ( cat config.log && false)" exited with 1.
$ make distdir VERSION=$HOST
make: *** No rule to make target `distdir'.  Stop.
```

**Comment by eklitzke** [comment]
This is cool, utACK fa7d9895ef3ec37ffeb6ca51333cf20bc8622e73 .

GCC 4.8 also supports -fsanitize=address and -fsanitize=thread (and newer GCC versions have a whole plethora or really interesting options). What do you think about using those options in Travis as well?

**Review by eklitzke — CHANGES_REQUESTED** [review]
There's a shell quoting issue with how travis is running this.

**Comment by eklitzke** [comment]
Here's an idea for another way to do this: https://github.com/eklitzke/bitcoin/commit/635e378383c41c8ef3ac03fca1755001c947b7f7 . The idea is to add -ftrapv when `--enable-debug` is used, and then use that option on all of the Travis jobs. `--enable-debug` also sets `-DDEBUG_LOCKORDER`.

What do you think of this approach?

**Comment by sipa** [comment]
@eklitzke There was some work earlier to get the code to run cleanly under sanitizers (see #9743 and #9512), though I don't think we ever made it perfectly runnable without warnings.

**Comment by eklitzke** [comment]
BTW the right way to fix the quoting issue in Travis is to use a bash array. It looks like Travis doesn't support these properly. If you still wanted to use this approach instead of modifying configure.ac as in my example two options:
 * `IFS` hacks
 * Create yet-another-Travis-variable and pass it down to the configure call

**Comment by practicalswift** [comment]
@eklitzke Thanks for reviewing. I've now cherry-picked your commit (which is the better solution) into this PR.

Please re-review :-)

**Comment by eklitzke** [comment]
It looks like this doesn't work for the MingW builds, I got timeouts in my travis runs that match the ones you just cherry-picked: https://travis-ci.org/eklitzke/bitcoin/builds/353587968 . Let's disable the option for those two builders.

I added some -fsanitize support to the configure script in #12692, which is a good start since it will make using those flags easier even if the code isn't currently clean under asan/tsan.

**Comment by practicalswift** [comment]
@eklitzke To keep the changes as small as possible I've now enabled `-ftrapv` only for one of the Travis jobs (job: "Qt4 & system libs"). Makes sense?

Please re-review :-)

**Comment by fanquake** [comment]
@practicalswift I don't think you've cherry-picked correctly, looks like you just swapped out the changes in your commit with @eklitzke's ?

**Comment by eklitzke** [comment]
This looks right to me, utACK e23dfbb01df044e6d0dc65f6b9333e6547202fa7.

**Comment by practicalswift** [comment]
@fanquake The cherry-pick was a previous version. e23dfbb01df044e6d0dc65f6b9333e6547202fa7 is correct. Please review :-)

**Comment by eklitzke** [comment]
FYI this is going to conflict with #12695 . Should be an easy merge conflict though.

**Comment by laanwj** [comment]
@theuni Can you take a look?

**Comment by practicalswift** [comment]
Rebased! Please re-review :-)

**Comment by practicalswift** [comment]
Judging from the Travis testing it seems that the trap is trigged indicating that we have a signed overflow taking place when running the tests. Let's find out where!

File: .travis.yml [hunk 1]
```diff
@@ -30,7 +30,7 @@ env:
 # 32-bit + dash
     - HOST=i686-pc-linux-gnu PACKAGES="g++-multilib python3-zmq" DEP_OPTS="NO_QT=1" RUN_TESTS=true GOAL="install" BITCOIN_CONFIG="--enable-zmq --enable-glibc-back-compat --enable-reduce-exports LDFLAGS=-static-libstdc++" USE_SHELL="/bin/dash"
 # x86_64 Linux (uses qt5 dev package instead of depends Qt to speed up build and avoid timeout)
-    - HOST=x86_64-unknown-linux-gnu PACKAGES="python3-zmq qtbase5-dev qttools5-dev-tools protobuf-compiler libdbus-1-dev libharfbuzz-dev" DEP_OPTS="NO_QT=1 NO_UPNP=1 DEBUG=1 ALLOW_HOST_PACKAGES=1" RUN_TESTS=true GOAL="install" BITCOIN_CONFIG="--enable-zmq --with-gui=qt5 --enable-glibc-back-compat --enable-reduce-exports CPPFLAGS=-DDEBUG_LOCKORDER"
```
Comment (theuni) [inline .travis.yml]: Removing ```DEBUG=1``` here should cause depends to no longer build as debug.

+1 on using --enable-debug, though.

**Comment by practicalswift** [comment]
Now using `AX_CHECK_COMPILE_FLAG` for checking `-trapv` availability when compiled under `--enable-debug`:

```
$ ./configure --enable-debug
…
checking whether C++ compiler accepts -ftrapv... yes
…
  CXXFLAGS      =  -Og -g3 -ftrapv  -Wstack-protector -fstack-protector-all   -g -O2 -Wall -Wextra -Wformat -Wvla -Wformat-security -Wno-unused-parameter
…
```

@theuni, please re-review :-)

**Comment by theuni** [comment]
Please add the CXXFLAG_WERROR argument so that we'll see false-positives. utACK otherwise.

**Comment by practicalswift** [comment]
@TheBlueMatt Added the `CXXFLAG_WERROR` argument. Please re-review :-)

**Comment by practicalswift** [comment]
Rebased!

**Comment by theuni** [comment]
utACK 8bf481fe2e99e6a0e16a6127d7df826dd4bfa2e9

**Comment by practicalswift** [comment]
Rebased!

**Comment by MarcoFalke** [comment]
ACK 98d842cb52 (that the rebase was done correctly, didn't look at anything else)

**Comment by sipa** [comment]
utACK 98d842cb52e69ea1f9961d908385c896e37fb877

**Pull request merged**